### PR TITLE
[Storage] [DataMovement] [Tests] Adding Directory Properties tests for Blob Directory and Share Directory

### PR DESCRIPTION
**Pull request by amnguye**
- Added Directory Properties tests for all blob types to Share files and vice versa and to other blob type directories
- Added directory test base for blob to share directory and vice versa to cut down on code duplication
- Ensured cancellation token is being passed on tests
- Updated download to local tests to not use public access

Bug Fixes:
- [DataMovement.Blobs] Fixed bug where using OAuth for Blob transfers would not preserve source properties to destination properties if enabled
- [DataMovement.Files.Shares] Fixed bug where copying a Azure Storage Blob to a Share File would not be able to convert Content Language and Content Encoding to the `string[]` type.
- [DataMovement.File.Shares] Fixed bug where LastWrittenOn property was not being preserved when copying a Share File to another Share File.

### CODE REVIEW

File: sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/DataMovementSharesExtensions.cs [hunk 1]
```diff
@@ -352,6 +367,22 @@ internal static partial class DataMovementSharesExtensions
                 lastModifiedTime: shareItem?.Properties?.LastModified,
                 properties: properties);
         }
+
+        private static string[] ConvertContentPropertyObjectToStringArray(string contentPropertyName, object contentPropertyValue)
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/DataMovementSharesExtensions.cs:371]: Didn't we already have this function somewhere?

File: sdk/storage/Azure.Storage.DataMovement/src/StorageResourceItemProperties.cs [hunk 2]
```diff
@@ -14,22 +14,22 @@ public class StorageResourceItemProperties
         /// <summary>
         /// The length of the resource.
         /// </summary>
-        public long? ResourceLength { get; }
+        public long? ResourceLength { get; set; }
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/src/StorageResourceItemProperties.cs:17]: Why do these need to be settable now?

**Review by jalauzon-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/DataMovementSharesExtensions.cs (hunk 1 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/DataMovementSharesExtensions.cs:371]: You're thinking of the other one where it does it for just `string` not `string[]`

File: sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/DataMovementSharesExtensions.cs [hunk 3]
```diff
@@ -264,6 +267,18 @@ internal static partial class DataMovementSharesExtensions
...
+            {
+                existingProperties.LastModifiedTime = fileProperties.LastModified;
+            }
+            if (existingProperties.ETag == default)
+            {
+                existingProperties.ETag = fileProperties.ETag;
+            }
+            if (existingProperties.ResourceLength == default)
+            {
+                existingProperties.ResourceLength = fileProperties.ContentLength;
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/DataMovementSharesExtensions.cs:280]: Made ResourceLength settable becuase of this change here.

There's a case where when we do directory transfers, the ShareFileItemProperties that comes back during enumeration does not have all the information we need. And then when we do, do a regular GetProperties call, this is where we ensure that information is updated. That way we do store the ResourceLength when we do get one.

File: sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/DataMovementSharesExtensions.cs (hunk 1 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/DataMovementSharesExtensions.cs:371]: see this I just made
https://github.com/Azure/azure-sdk-for-net/pull/45398/files#r1712077627

So I had to make this change cause the length wasn't being stored correctly, therefore not correctly seeing that there is a length in fact for downloading the file.

File: sdk/storage/Azure.Storage.DataMovement/src/StorageResourceItemProperties.cs (hunk 2 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/src/StorageResourceItemProperties.cs:17]: see this I just made
https://github.com/Azure/azure-sdk-for-net/pull/45398/files#r1712077627

So I had to make this change cause the length wasn't being stored correctly, therefore not correctly seeing that there is a length in fact for downloading the file.

**Pull request merged**